File: backend/app/services/security/file_validator.py
"""
File Validation Service - TuCitaSegura

Comprehensive file validation including:
- File size validation
- MIME type verification (not just extension)
- Format whitelisting
- Security checks against malicious files
- Image validation
"""

import magic
import mimetypes
from typing import Optional, List, Dict, Any, BinaryIO
from pathlib import Path
from io import BytesIO
import logging
from PIL import Image
from fastapi import UploadFile, HTTPException
from dataclasses import dataclass

from app.core.config import settings

logger = logging.getLogger(__name__)


@dataclass
class FileValidationResult:
    """Result of file validation"""
    is_valid: bool
    mime_type: str
    extension: str
    size_bytes: int
    errors: List[str]
    warnings: List[str]
    metadata: Dict[str, Any]


class FileValidator:
    """
    Secure file validation service

    Features:
    - MIME type detection (real content, not extension)
    - Size validation
    - Format whitelisting
    - Security checks
    - Image validation (dimensions, format)
    """

    # Allowed MIME types for different file categories
    ALLOWED_IMAGE_TYPES = {
        'image/jpeg',
        'image/jpg',
        'image/png',
        'image/webp',
        'image/gif'
    }

    ALLOWED_DOCUMENT_TYPES = {
        'application/pdf',
        'application/msword',
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        'text/plain'
    }

    # Dangerous MIME types (never allow)
    DANGEROUS_TYPES = {
        'application/x-executable',
        'application/x-dosexec',
        'application/x-mach-binary',
        'application/x-sh',
        'application/x-shellscript',
        'text/x-script.python',
        'text/x-php',
        'application/javascript',
        'text/javascript',
        'application/x-httpd-php'
    }

    # Suspicious file extensions
    DANGEROUS_EXTENSIONS = {
        '.exe', '.bat', '.cmd', '.sh', '.app', '.deb', '.rpm',
        '.msi', '.dmg', '.pkg', '.run', '.bin', '.com', '.scr',
        '.vbs', '.js', '.jar', '.apk', '.ipa', '.py', '.php',
        '.asp', '.aspx', '.jsp', '.cgi'
    }

    def __init__(self):
        """Initialize file validator"""
        self.max_image_size = settings.CV_MAX_IMAGE_SIZE  # 5MB
        self.max_document_size = 10 * 1024 * 1024  # 10MB

        # Parse allowed formats from settings
        self.allowed_image_formats = {
            fmt.strip().lower()
            for fmt in settings.CV_ALLOWED_FORMATS.split(',')
        }

        logger.info(f"FileValidator initialized with max image size: {self.max_image_size}")

    async def validate_upload_file(
        self,
        file: UploadFile,
        category: str = 'image',
        max_size: Optional[int] = None
    ) -> FileValidationResult:
        """
        Validate an uploaded file

        Args:
            file: FastAPI UploadFile object
            category: File category ('image', 'document')
            max_size: Optional custom max size in bytes

        Returns:
            FileValidationResult

        Raises:
            HTTPException: If file is invalid
        """
        errors = []
        warnings = []
        metadata = {}

        # Read file content
        try:
            content = await file.read()
            await file.seek(0)  # Reset file pointer
        except Exception as e:
            logger.error(f"Error reading file: {e}")
            raise HTTPException(status_code=400, detail="Could not read file")

        # Validate size
        size_bytes = len(content)
        metadata['size_bytes'] = size_bytes
        metadata['size_mb'] = round(size_bytes / (1024 * 1024), 2)

        max_allowed_size = max_size or (
            self.max_image_size if category == 'image' else self.max_document_size
        )

        if size_bytes > max_allowed_size:
            errors.append(
                f"File too large: {metadata['size_mb']}MB "
                f"(max: {max_allowed_size / (1024 * 1024)}MB)"
            )

        if size_bytes == 0:
            errors.append("File is empty")

        # Detect real MIME type from content
        try:
            mime = magic.from_buffer(content, mime=True)
            metadata['detected_mime'] = mime
        except Exception as e:
            logger.error(f"Error detecting MIME type: {e}")
            errors.append("Could not detect file type")
            mime = 'application/octet-stream'

        # Get extension from filename
        file_path = Path(file.filename) if file.filename else Path('unknown')
        extension = file_path.suffix.lower()
        metadata['extension'] = extension
        metadata['filename'] = file.filename

        # Check for dangerous extensions
        if extension in self.DANGEROUS_EXTENSIONS:
            errors.append(f"Dangerous file extension: {extension}")

        # Check for dangerous MIME types
        if mime in self.DANGEROUS_TYPES:
            errors.append(f"Dangerous file type detected: {mime}")

        # Validate based on category
        if category == 'image':
            image_errors, image_warnings, image_metadata = self._validate_image(
                content, mime, extension
            )
            errors.extend(image_errors)
            warnings.extend(image_warnings)
            metadata.update(image_metadata)
        elif category == 'document':
            doc_errors, doc_metadata = self._validate_document(
                content, mime, extension
            )
            errors.extend(doc_errors)
            metadata.update(doc_metadata)

        # Check MIME type vs extension mismatch
        expected_mime = mimetypes.guess_type(file.filename)[0] if file.filename else None
        if expected_mime and mime != expected_mime:
            warnings.append(
                f"MIME type mismatch: extension suggests {expected_mime}, "
                f"but content is {mime}"
            )

        is_valid = len(errors) == 0

        return FileValidationResult(
            is_valid=is_valid,
            mime_type=mime,
            extension=extension,
            size_bytes=size_bytes,
            errors=errors,
            warnings=warnings,
            metadata=metadata
        )

    def _validate_image(
        self,
        content: bytes,
        mime: str,
        extension: str
    ) -> tuple[List[str], List[str], Dict[str, Any]]:
        """Validate image file"""
        errors = []
        warnings = []
        metadata = {}

        # Check MIME type
        if mime not in self.ALLOWED_IMAGE_TYPES:
            errors.append(
                f"Invalid image type: {mime}. "
                f"Allowed: {', '.join(self.ALLOWED_IMAGE_TYPES)}"
            )

        # Check extension
        ext_without_dot = extension.lstrip('.')
        if ext_without_dot not in self.allowed_image_formats:
            errors.append(
                f"Invalid image format: {extension}. "
                f"Allowed: {', '.join(self.allowed_image_formats)}"
            )

        # Validate with PIL
        try:
            img = Image.open(BytesIO(content))
            metadata['width'] = img.width
            metadata['height'] = img.height
            metadata['format'] = img.format
            metadata['mode'] = img.mode

            # Validate dimensions
            if img.width < 100 or img.height < 100:
                warnings.append(
                    f"Image too small: {img.width}x{img.height} "
                    "(minimum recommended: 100x100)"
                )

            if img.width > 8000 or img.height > 8000:
                warnings.append(
                    f"Image very large: {img.width}x{img.height} "
                    "(may cause performance issues)"
                )

            # Validate aspect ratio (prevent extreme ratios)
            aspect_ratio = img.width / img.height
            if aspect_ratio > 5 or aspect_ratio < 0.2:
                warnings.append(
                    f"Unusual aspect ratio: {aspect_ratio:.2f} "
                    "(image may be distorted)"
                )

            # Verify image can be loaded (detect corrupted files)
            img.verify()

        except Exception as e:
            errors.append(f"Invalid or corrupted image: {str(e)}")
            logger.error(f"Image validation error: {e}")

        return errors, warnings, metadata

    def _validate_document(
        self,
        content: bytes,
        mime: str,
        extension: str
    ) -> tuple[List[str], Dict[str, Any]]:
        """Validate document file"""
        errors = []
        metadata = {}

        # Check MIME type
        if mime not in self.ALLOWED_DOCUMENT_TYPES:
            errors.append(
                f"Invalid document type: {mime}. "
                f"Allowed: {', '.join(self.ALLOWED_DOCUMENT_TYPES)}"
            )

        # Check for embedded scripts (basic check)
        if b'<script' in content.lower() or b'javascript:' in content.lower():
            errors.append("Document contains potentially malicious scripts")

        return errors, metadata

    def validate_file_sync(
        self,
        file_path: str,
        category: str = 'image'
    ) -> FileValidationResult:
        """
        Validate a file from filesystem (synchronous)

        Args:
            file_path: Path to file
            category: File category

        Returns:
            FileValidationResult
        """
        errors = []
        warnings = []
        metadata = {}

        path = Path(file_path)

        # Check if file exists
        if not path.exists():
            errors.append(f"File not found: {file_path}")
            return FileValidationResult(
                is_valid=False,
                mime_type='',
                extension='',
                size_bytes=0,
                errors=errors,
                warnings=warnings,
                metadata=metadata
            )

        # Get file size
        size_bytes = path.stat().st_size
        metadata['size_bytes'] = size_bytes

        # Read file
        try:
            with open(file_path, 'rb') as f:
                content = f.read()
        except Exception as e:
            errors.append(f"Could not read file: {e}")
            return FileValidationResult(
                is_valid=False,
                mime_type='',
                extension='',
                size_bytes=size_bytes,
                errors=errors,
                warnings=warnings,
                metadata=metadata
            )

        # Detect MIME type
        try:
            mime = magic.from_file(file_path, mime=True)
            metadata['detected_mime'] = mime
        except Exception as e:
            logger.error(f"Error detecting MIME type: {e}")
            errors.append("Could not detect file type")
            mime = 'application/octet-stream'

        extension = path.suffix.lower()
        metadata['extension'] = extension

        # Validate based on category
        if category == 'image':
            image_errors, image_warnings, image_metadata = self._validate_image(
                content, mime, extension
            )
            errors.extend(image_errors)
            warnings.extend(image_warnings)
            metadata.update(image_metadata)

        is_valid = len(errors) == 0

        return FileValidationResult(
            is_valid=is_valid,
            mime_type=mime,
            extension=extension,
            size_bytes=size_bytes,
            errors=errors,
            warnings=warnings,
            metadata=metadata
        )


# Global instance
file_validator = FileValidator()

File: webapp/js/file-validator.js
/**
 * File Validator - Frontend
 *
 * Client-side file validation before uploading to Firebase Storage
 * Provides security checks, MIME type validation, and size limits
 */

import { logger } from './logger.js';

// Allowed MIME types for images
const ALLOWED_IMAGE_TYPES = new Set([
  'image/jpeg',
  'image/jpg',
  'image/png',
  'image/webp',
  'image/gif'
]);

// Dangerous file extensions (never allow)
const DANGEROUS_EXTENSIONS = new Set([
  '.exe', '.bat', '.cmd', '.sh', '.app', '.deb', '.rpm',
  '.msi', '.dmg', '.pkg', '.run', '.bin', '.com', '.scr',
  '.vbs', '.js', '.jar', '.apk', '.ipa', '.py', '.php',
  '.asp', '.aspx', '.jsp', '.cgi', '.pl', '.rb'
]);

// Max file sizes
const MAX_IMAGE_SIZE = 5 * 1024 * 1024; // 5MB
const MAX_DOCUMENT_SIZE = 10 * 1024 * 1024; // 10MB

/**
 * Validation result object
 */
export class FileValidationResult {
  constructor(isValid, errors = [], warnings = [], metadata = {}) {
    this.isValid = isValid;
    this.errors = errors;
    this.warnings = warnings;
    this.metadata = metadata;
  }
}

/**
 * File Validator class
 */
export class FileValidator {
  constructor() {
    this.maxImageSize = MAX_IMAGE_SIZE;
    this.maxDocumentSize = MAX_DOCUMENT_SIZE;
  }

  /**
   * Validate an image file
   * @param {File} file - File object to validate
   * @returns {Promise<FileValidationResult>}
   */
  async validateImage(file) {
    const errors = [];
    const warnings = [];
    const metadata = {};

    try {
      // Basic checks
      if (!file) {
        errors.push('No file provided');
        return new FileValidationResult(false, errors, warnings, metadata);
      }

      // File size
      metadata.sizeBytes = file.size;
      metadata.sizeMB = (file.size / (1024 * 1024)).toFixed(2);

      if (file.size === 0) {
        errors.push('File is empty');
      }

      if (file.size > this.maxImageSize) {
        errors.push(`File too large: ${metadata.sizeMB}MB (max: ${(this.maxImageSize / (1024 * 1024)).toFixed(0)}MB)`);
      }

      // File name and extension
      const fileName = file.name || 'unknown';
      const extension = this.getFileExtension(fileName);
      metadata.fileName = fileName;
      metadata.extension = extension;

      // Check dangerous extensions
      if (DANGEROUS_EXTENSIONS.has(extension.toLowerCase())) {
        errors.push(`Dangerous file extension: ${extension}`);
      }

      // MIME type validation
      const mimeType = file.type || '';
      metadata.mimeType = mimeType;

      if (!mimeType) {
        errors.push('Could not detect file type');
      } else if (!ALLOWED_IMAGE_TYPES.has(mimeType.toLowerCase())) {
        errors.push(`Invalid image type: ${mimeType}. Allowed: ${Array.from(ALLOWED_IMAGE_TYPES).join(', ')}`);
      }

      // Validate extension matches MIME type
      const expectedExtensions = {
        'image/jpeg': ['.jpg', '.jpeg'],
        'image/png': ['.png'],
        'image/webp': ['.webp'],
        'image/gif': ['.gif']
      };

      if (mimeType && expectedExtensions[mimeType]) {
        const allowed = expectedExtensions[mimeType];
        if (!allowed.includes(extension.toLowerCase())) {
          warnings.push(`Extension ${extension} doesn't match MIME type ${mimeType}`);
        }
      }

      // Load and validate image
      if (errors.length === 0) {
        try {
          const imageData = await this.loadImage(file);
          metadata.width = imageData.width;
          metadata.height = imageData.height;

          // Dimension checks
          if (imageData.width < 100 || imageData.height < 100) {
            warnings.push(`Image too small: ${imageData.width}x${imageData.height} (minimum recommended: 100x100)`);
          }

          if (imageData.width > 8000 || imageData.height > 8000) {
            warnings.push(`Image very large: ${imageData.width}x${imageData.height} (may cause performance issues)`);
          }

          // Aspect ratio
          const aspectRatio = imageData.width / imageData.height;
          metadata.aspectRatio = aspectRatio.toFixed(2);

          if (aspectRatio > 5 || aspectRatio < 0.2) {
            warnings.push(`Unusual aspect ratio: ${aspectRatio.toFixed(2)} (image may be distorted)`);
          }

        } catch (error) {
          errors.push(`Invalid or corrupted image: ${error.message}`);
          logger.error('Image validation error', error, { fileName });
        }
      }

      const isValid = errors.length === 0;

      logger.debug('File validation completed', {
        fileName,
        isValid,
        errorCount: errors.length,
        warningCount: warnings.length
      });

      return new FileValidationResult(isValid, errors, warnings, metadata);

    } catch (error) {
      logger.error('File validation error', error);
      errors.push(`Validation error: ${error.message}`);
      return new FileValidationResult(false, errors, warnings, metadata);
    }
  }

  /**
   * Load and validate image file
   * @param {File} file
   * @returns {Promise<{width: number, height: number, img: HTMLImageElement}>}
   */
  loadImage(file) {
    return new Promise((resolve, reject) => {
      const img = new Image();
      const url = URL.createObjectURL(file);

      img.onload = () => {
        URL.revokeObjectURL(url);
        resolve({
          width: img.naturalWidth,
          height: img.naturalHeight,
          img
        });
      };

      img.onerror = () => {
        URL.revokeObjectURL(url);
        reject(new Error('Failed to load image'));
      };

      img.src = url;
    });
  }

  /**
   * Get file extension from filename
   * @param {string} fileName
   * @returns {string}
   */
  getFileExtension(fileName) {
    const parts = fileName.split('.');
    if (parts.length > 1) {
      return '.' + parts[parts.length - 1];
    }
    return '';
  }

  /**
   * Validate multiple files
   * @param {FileList|File[]} files
   * @returns {Promise<FileValidationResult[]>}
   */
  async validateImages(files) {
    const results = [];

    for (const file of files) {
      const result = await this.validateImage(file);
      results.push(result);
    }

    return results;
  }

  /**
   * Check if file type is allowed
   * @param {string} mimeType
   * @returns {boolean}
   */
  isAllowedImageType(mimeType) {
    return ALLOWED_IMAGE_TYPES.has(mimeType.toLowerCase());
  }

  /**
   * Check if extension is dangerous
   * @param {string} extension
   * @returns {boolean}
   */
  isDangerousExtension(extension) {
    return DANGEROUS_EXTENSIONS.has(extension.toLowerCase());
  }

  /**
   * Format file size in human-readable format
   * @param {number} bytes
   * @returns {string}
   */
  formatFileSize(bytes) {
    if (bytes === 0) return '0 Bytes';

    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
  }
}

/**
 * Helper function to show validation errors to user
 * @param {FileValidationResult} result
 * @param {Function} showToast - Toast notification function
 */
export function showValidationErrors(result, showToast) {
  if (!result.isValid) {
    result.errors.forEach(error => {
      showToast(error, 'error');
      logger.error('File validation error', { error });
    });
  }

  result.warnings.forEach(warning => {
    showToast(warning, 'warning');
    logger.warn('File validation warning', { warning });
  });
}

/**
 * Create upload preview with validation
 * @param {File} file
 * @param {FileValidationResult} validation
 * @returns {HTMLElement}
 */
export function createUploadPreview(file, validation) {
  const container = document.createElement('div');
  container.className = 'upload-preview';

  const validator = new FileValidator();

  // Preview image
  const preview = document.createElement('img');
  preview.src = URL.createObjectURL(file);
  preview.onload = () => URL.revokeObjectURL(preview.src);
  preview.className = 'preview-image';
  container.appendChild(preview);

  // File info
  const info = document.createElement('div');
  info.className = 'preview-info';

  const fileName = document.createElement('div');
  fileName.className = 'file-name';
  fileName.textContent = file.name;
  info.appendChild(fileName);

  const fileSize = document.createElement('div');
  fileSize.className = 'file-size';
  fileSize.textContent = validator.formatFileSize(file.size);
  info.appendChild(fileSize);

  if (validation.metadata.width && validation.metadata.height) {
    const dimensions = document.createElement('div');
    dimensions.className = 'file-dimensions';
    dimensions.textContent = `${validation.metadata.width} × ${validation.metadata.height}`;
    info.appendChild(dimensions);
  }

  container.appendChild(info);

  // Validation status
  const status = document.createElement('div');
  status.className = `validation-status ${validation.isValid ? 'valid' : 'invalid'}`;

  if (validation.isValid) {
    status.innerHTML = '<i class="fas fa-check-circle"></i> Válido';
    if (validation.warnings.length > 0) {
      status.innerHTML += ` <span class="warning-count">(${validation.warnings.length} avisos)</span>`;
    }
  } else {
    status.innerHTML = '<i class="fas fa-times-circle"></i> Inválido';
  }

  container.appendChild(status);

  // Errors and warnings
  if (validation.errors.length > 0 || validation.warnings.length > 0) {
    const messages = document.createElement('div');
    messages.className = 'validation-messages';

    validation.errors.forEach(error => {
      const errorDiv = document.createElement('div');
      errorDiv.className = 'validation-error';
      errorDiv.innerHTML = `<i class="fas fa-exclamation-circle"></i> ${error}`;
      messages.appendChild(errorDiv);
    });

    validation.warnings.forEach(warning => {
      const warningDiv = document.createElement('div');
      warningDiv.className = 'validation-warning';
      warningDiv.innerHTML = `<i class="fas fa-exclamation-triangle"></i> ${warning}`;
      messages.appendChild(warningDiv);
    });

    container.appendChild(messages);
  }

  return container;
}

// Global instance
export const fileValidator = new FileValidator();

// Default export
export default fileValidator;
